### PR TITLE
Refactor cmake to build either static or shared

### PR DESCRIPTION
**Pull request by nafur**
This PR simplifies the cmake setup to only build either shared or statically. It also attempts to fix windows builds, both shared and static.

### CODE REVIEW

File: cmake/FindPoly.cmake [hunk 1]
```diff
@@ -138,41 +132,32 @@ endif()
...
+else()
+  add_library(Poly SHARED IMPORTED GLOBAL)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    set_target_properties(Poly PROPERTIES IMPORTED_IMPLIB "${Poly_LIBRARIES}")
+  endif()
+
+  add_library(Polyxx SHARED IMPORTED GLOBAL)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    set_target_properties(Polyxx PROPERTIES IMPORTED_IMPLIB "${PolyXX_LIBRARIES}")
+  endif()
```
Comment (mpreiner) [inline cmake/FindPoly.cmake:148]:
```suggestion
  add_library(Poly SHARED IMPORTED GLOBAL)
  add_library(Polyxx SHARED IMPORTED GLOBAL)
  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set_target_properties(Poly PROPERTIES IMPORTED_IMPLIB "${Poly_LIBRARIES}")
    set_target_properties(Polyxx PROPERTIES IMPORTED_IMPLIB "${PolyXX_LIBRARIES}")
  endif()
```

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -105,6 +105,7 @@ if(TARGET cvc5::cvc5jar)
 endif()
 
 if(CVC5_BINDINGS_PYTHON)
+  message(STATUS "with python examples")
```
Comment (mpreiner) [inline examples/CMakeLists.txt:108]:
```suggestion
  message(STATUS "with Python examples")
```

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -1324,53 +1324,42 @@ add_dependencies(cvc5-obj cvc5base cvc5context)
 # Add libcvc5 dependencies for generated sources.
 add_dependencies(cvc5-obj gen-expr gen-versioninfo gen-options gen-tags gen-theory)
 
+if(ENABLE_STATIC_BUILD)
+  add_library(cvc5 STATIC $<TARGET_OBJECTS:cvc5-obj> $<TARGET_OBJECTS:cvc5base> $<TARGET_OBJECTS:cvc5context>)
+else()
+  add_library(cvc5 SHARED $<TARGET_OBJECTS:cvc5-obj> $<TARGET_OBJECTS:cvc5base> $<TARGET_OBJECTS:cvc5context>)
+endif()
```
Comment (mpreiner) [inline src/CMakeLists.txt:1331]: I'm currently wondering why we don't use `BUILD_SHARED_LIBS`? If we would we wouldn't need this if/else and
```
add_library(cvc5 $<TARGET_OBJECTS:cvc5-obj> $<TARGET_OBJECTS:cvc5base> $<TARGET_OBJECTS:cvc5context>)
```
would suffice.

File: src/api/java/CMakeLists.txt [hunk 4]
```diff
@@ -109,22 +109,41 @@ message(STATUS "JAVA_INCLUDE_PATH    : ${JAVA_INCLUDE_PATH}")
...
+    jni/grammar.cpp
+    jni/op.cpp
+    jni/option_info.cpp
+    jni/result.cpp
+    jni/solver.cpp
+    jni/sort.cpp
+    jni/stat.cpp
+    jni/statistics.cpp
+    jni/term.cpp)
+endif()
```
Comment (mpreiner) [inline src/api/java/CMakeLists.txt:146]: Ditto.

File: src/parser/CMakeLists.txt [hunk 5]
```diff
@@ -114,27 +114,21 @@ target_compile_definitions(cvc5parser-objs PUBLIC -D__BUILDING_CVC5PARSERLIB -Dc
...
 
-add_library(cvc5parser-shared SHARED $<TARGET_OBJECTS:cvc5parser-objs>)
-set_target_properties(cvc5parser-shared PROPERTIES SOVERSION ${CVC5_SOVERSION})
-set_target_properties(cvc5parser-shared PROPERTIES OUTPUT_NAME cvc5parser)
-target_link_libraries(cvc5parser-shared PRIVATE cvc5-shared)
-target_link_libraries(cvc5parser-shared PRIVATE ANTLR3)
+if(ENABLE_STATIC_BUILD)
+  add_library(cvc5parser STATIC $<TARGET_OBJECTS:cvc5parser-objs>)
+else()
+  add_library(cvc5parser SHARED $<TARGET_OBJECTS:cvc5parser-objs>)
```
Comment (mpreiner) [inline src/parser/CMakeLists.txt:120]: Ditto.

File: src/CMakeLists.txt (hunk 3 above)
Comment (nafur) [reply src/CMakeLists.txt:1331]: Oh. I was not aware of this mechanism. Yes, we should use it!

Comment (nafur) [reply src/CMakeLists.txt]: Okay, done.

File: cmake/FindGMP.cmake [hunk 6]
```diff
@@ -62,83 +52,47 @@ if(NOT GMP_FOUND_SYSTEM)
...
-set_target_properties(GMP_SHARED PROPERTIES
+
+if(BUILD_SHARED_LIBS)
+  add_library(GMP SHARED IMPORTED GLOBAL)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    set_target_properties(GMP PROPERTIES IMPORTED_IMPLIB "${GMP_LIBRARIES}")
+  endif()
+else()
+  add_library(GMP STATIC IMPORTED GLOBAL)
+endif()
```
Comment (mpreiner) [inline cmake/FindGMP.cmake:91]:
```suggestion
add_library(GMP IMPORTED GLOBAL)
if(BUILD_SHARED_LIBS)
  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set_target_properties(GMP PROPERTIES IMPORTED_IMPLIB "${GMP_LIBRARIES}")
  endif()
endif()
```

File: cmake/FindPoly.cmake [hunk 7]
```diff
@@ -138,41 +132,29 @@ endif()
...
+  add_library(Poly SHARED IMPORTED GLOBAL)
+  add_library(Polyxx SHARED IMPORTED GLOBAL)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    set_target_properties(Poly PROPERTIES IMPORTED_IMPLIB "${Poly_LIBRARIES}")
+    set_target_properties(Polyxx PROPERTIES IMPORTED_IMPLIB "${PolyXX_LIBRARIES}")
+  endif()
+else()
+  add_library(Poly STATIC IMPORTED GLOBAL)
+  add_library(Polyxx STATIC IMPORTED GLOBAL)
+endif()
```
Comment (mpreiner) [inline cmake/FindPoly.cmake:146]:
```suggestion
add_library(Poly IMPORTED GLOBAL)
add_library(Polyxx IMPORTED GLOBAL)
if(BUILD_SHARED_LIBS)
  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set_target_properties(Poly PROPERTIES IMPORTED_IMPLIB "${Poly_LIBRARIES}")
    set_target_properties(Polyxx PROPERTIES IMPORTED_IMPLIB "${PolyXX_LIBRARIES}")
  endif()
endif()
```

File: cmake/FindCLN.cmake [hunk 8]
```diff
@@ -77,31 +68,28 @@ if(NOT CLN_FOUND_SYSTEM)
...
 
 set(CLN_FOUND TRUE)
 
-add_library(CLN_SHARED SHARED IMPORTED GLOBAL)
-set_target_properties(CLN_SHARED PROPERTIES
+if(BUILD_SHARED_LIBS)
+  add_library(CLN SHARED IMPORTED GLOBAL)
+else()
+  add_library(CLN STATIC IMPORTED GLOBAL)
+endif()
```
[the remark this answers is not in the document]
Comment (nafur) [reply cmake/FindCLN.cmake:87]: Yes, but unfortunately not when using `IMPORTED`... (also see the last few commits)

File: cmake/FindGMP.cmake (hunk 6 above)
Comment (nafur) [reply cmake/FindGMP.cmake:91]: ditto.

File: cmake/FindPoly.cmake (hunk 7 above)
Comment (nafur) [reply cmake/FindPoly.cmake:146]: ditto.

File: CMakeLists.txt [hunk 9]
```diff
@@ -268,6 +268,20 @@ enable_testing()
...
+  else()
+    set(CMAKE_FIND_LIBRARY_SUFFIXES .so .dylib .a)
+  endif()
+else()
+  if (WIN32)
+    set(CMAKE_FIND_LIBRARY_SUFFIXES .lib .a .dll)
+  else()
+    set(CMAKE_FIND_LIBRARY_SUFFIXES .a .so .dylib)
+  endif()
+endif()
```
[the remark this answers is not in the document]
Comment (nafur) [reply CMakeLists.txt:283]: No, setting `BUILD_SHARED_LIBS` does not seem to change `CMAKE_FIND_LIBRARY_SUFFIXES`.

**Review by mpreiner — APPROVED** [review]

**Pull request merged**